File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.ts
const [N, ...arr]: number[] = require("fs").readFileSync(process.platform==="linux"?0:"input.txt").toString().trim().split("\n").map((x: string)=>parseInt(x));
const stack: number[] = [];
const cnt: number[] = [];
let l = 0; // stack.length
let ans = 0;
let tmp = 0;

for (let i=0; i<N; i++) {
  while (l > 0 && arr[stack[l-1]] < arr[i]) {
    stack.pop();
    tmp = cnt.pop() as number;
    ans += (l > 1 ? tmp * 2: tmp) + (tmp * (tmp-1)) / 2;
    l -= 1;
  }
  if (l > 0 && arr[stack[l-1]] === arr[i]) {
    cnt[l-1] += 1;
  } else {
    stack.push(i);
    cnt.push(1);
    l += 1;
  }
}

while (l > 0) {
  tmp = cnt.pop() as number;
  ans += (l > 1 ? tmp : 0) + (tmp * (tmp-1)) / 2;
  l -= 1;
}

console.log(ans);

File: 백준/Platinum/3015. 오아시스 재결합/오아시스 재결합.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N];
        int[] stack = new int[N];
        int[] cnt = new int[N];
        int t = -1;
        long ans = 0;

        for (int i = 0; i < N; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }

        for (int i = 0; i < N; i++) {
            while (t > -1 && arr[stack[t]] < arr[i]) {
                long tmp = cnt[t--];
                if (t > -1) {
                    ans += (tmp * (tmp + 3)) / 2;
                } else {
                    ans += (tmp * (tmp + 1)) / 2;
                }
            }

            if (t > -1 && arr[stack[t]] == arr[i]) {
                cnt[t]++;
            } else {
                t++;
                stack[t] = i;
                cnt[t] = 1;
            }
        }

        while (t > -1) {
            long tmp = cnt[t--];
            if (t > -1) {
                ans += (tmp * (tmp + 1)) / 2;
            } else {
                ans += (tmp * (tmp - 1)) / 2;
            }
        }

        System.out.println(ans);
    }
}
